source/initial_composition/porosity.cc

```diff
@@ -0,0 +1,95 @@
+                             "compositional field called 'porosity' to initialize. Please add a "
+                             "compositional field with this name."));
+
+      const MaterialModel::MeltFractionModel<dim> *material_model =
+        dynamic_cast<const MaterialModel::MeltFractionModel<dim>* > (&this->get_material_model());
+      AssertThrow(material_model != NULL,
+                  ExcMessage("The used material model is not derived from the 'MeltFractionModel' class, "
+                             "and therefore does not support computing equilibrium melt fractions. "
+                             "This is incompatible with the 'porosity' "
+                             "initial composition plugin, which needs to compute these melt fractions."));
+
+      const unsigned int porosity_index = this->introspection().compositional_index_for_name("porosity");
+      if (compositional_index == porosity_index)
+        {
+          MaterialModel::MaterialModelInputs<dim> in(1, this->n_compositional_fields());
+
+          in.position[0] = position;
+          in.temperature[0] = this->get_initial_temperature_manager().initial_temperature(position);
+          in.pressure[0] = this->get_adiabatic_conditions().pressure(position);
+          in.pressure_gradient[0] = 0.0;
+          in.velocity[0] = 0.0;
+
+          // Use the initial composition, except for the porosity, to prevent
+          // infinite recursion
+          for (unsigned int i = 0; i < this->n_compositional_fields(); ++i)
+            if (i != porosity_index)
+              in.composition[0][i] = this->get_initial_composition_manager().initial_composition(position,i);
+            else
+              in.composition[0][i] = 0.0;
+
+          in.strain_rate[0] = SymmetricTensor<2,dim>();
+          in.cell= NULL;
+
+          std::vector<double> melt_fraction(1);
+          material_model->melt_fractions(in,melt_fraction);
+          return melt_fraction[0];
+        }
+      return 0.0;
+    }
+  }
+}
+
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialComposition
+  {
+    ASPECT_REGISTER_INITIAL_COMPOSITION_MODEL(Porosity,
+                                              "porosity",
+                                              "A class that implements initial conditions for the porosity field "
+                                              "by computing the equilibrium melt fraction for the given initial "
+                                              "condition and reference pressure profile. Note that this plugin only "
+                                              "works if there is a compositional field called 'porosity', and the "
+                                              "used material model implements the 'MeltFractionModel' interface. "
+                                              "All compositional fields except porosity are not changed by this plugin.")
```

Comment:
Maybe it is more appropriate to say that it returns 0 for other fields? (or in addition)?